File: Sqrt(x).cpp
class Solution {
public:
    /**
     * @param x: An integer
     * @return: The sqrt of x
     */
    int sqrt(int x) {
        // write your code here
        //solution 1
        // int s = 1;
        // int e = x / 2;
        // while (s <= e){
        //     int m = s + (e - s) / 2;
        //     if ( m > x / m){
        //         e = m - 1;
        //     }else{
        //         s = m + 1;
        //     }
        // }
        // return e;//s - 1 is also ok.
        //solution 2
        // int s = 1;
        // int e = x / 2 + 1;
        // while (s < e){
        //     int m = s + (e - s) / 2;
        //     if (m > x / m){
        //         e = m;
        //     }else{
        //         s = m + 1;
        //     }
        // }
        // return s - 1;//e - 1 is also ok, e == s.
        int s = 1;
        int e = x / 2 + 1;
        while (e - s > 1){
            int m = s + (e - s) / 2;
            if (m > x / m){
                e = m;
            }else {
                s = m;
            }
        }
        return s;// e - 1 is ok.
    }
};

File: Sqrt(x).py
class Solution:
    """
    @param x: An integer
    @return: The sqrt of x
    """
    def sqrt(self, x):
        # write your code here
        
        left, right = 0, x
        while left + 1 < right:
            mid = left + (right - left) / 2
            if mid * mid <= x:
                left = mid
            else:
                right = mid
                
        #error: we should not check the left because there is no exception for the left.(0 is belong to the left * left <= x)
        # if left * left <= x:
        #     return left
        if right * right <= x:
            return right
        else:
            return left
                